src/Combinatorics/EnumerativeCombinatorics/compositions.jl

```diff
@@ -0,0 +1,376 @@
+function number_of_compositions(n::IntegerUnion, k::IntegerUnion)
+  if n < 0 || k < 0
+    return ZZ(0)
+  end
+
+  if n == 0
+    if k == 0
+      return ZZ(1)
+    else
+      return ZZ(0)
+    end
+  else
+    return binomial(ZZ(n - 1), ZZ(k - 1))
+  end
+end
+
+@doc raw"""
+    number_of_compositions(n::IntegerUnion)
+
+Return the number of compositions of the non-negative integer `n`.
+For `n < 0`, return `0`.
+"""
+function number_of_compositions(n::IntegerUnion)
+  if n < 0
+    return ZZ(0)
+  elseif n == 0
+    return ZZ(1)
+  else
+    return ZZ(2)^(n - 1)
+  end
+end
+
+################################################################################
+#
+#  Generating compositions with fixed number of parts
+#
+################################################################################
+
+@doc raw"""
+    compositions(n::IntegerUnion, k::IntegerUnion)
+
+Return an iterator over all compositions of a non-negative integer `n` into
+`k` parts, produced in lexicographically *descending* order.
+
+By a composition of `n` into `k` parts we mean a sequence of `k` positive integers
+whose sum is `n`.
+
+# Examples
+```jldoctest
+julia> C = compositions(4, 2)
+Iterator over the compositions of 4 into 2 parts
+
+julia> collect(C)
+3-element Vector{Oscar.Composition{Int64}}:
+ [3, 1]
+ [2, 2]
+ [1, 3]
+```
+"""
+function compositions(n::IntegerUnion, k::IntegerUnion)
+  kk = Int(k)
+  return CompositionsFixedNumParts(n, kk)
+end
+
+base(C::CompositionsFixedNumParts) = C.n
+parts(C::CompositionsFixedNumParts) = C.k
+
+Base.eltype(C::CompositionsFixedNumParts{T}) where T = Composition{T}
+
+function Base.show(io::IO, C::CompositionsFixedNumParts)
+  if get(io, :supercompact, false)
+    print(io, "Iterator")
+  else
+    io = pretty(io)
+    print(io, "Iterator over the compositions of $(base(C)) into ", ItemQuantity(parts(C), "part"))
+  end
+end
+
+Base.length(C::CompositionsFixedNumParts) = BigInt(number_of_compositions(base(C), parts(C)))
+
+function Base.iterate(C::CompositionsFixedNumParts{T}, state::Union{Nothing,Vector{T}}=nothing) where T
+  w = iterate(C.weak_comp_iter, state)
+  if w === nothing
+    return nothing
+  end
+
+  # w[1] is a weak composition of n - k into k parts, we get a composition of n
+  # into k parts by adding 1 to every entry.
+  c = data(w[1])
+  s = w[2]
+  for i in 1:length(c)
+    c[i] += 1
+  end
+
+  return composition(c, check=false), s
+end
+
+################################################################################
+#
+#  Generating all compositions
+#
+################################################################################
+
+@doc raw"""
+    compositions(n::IntegerUnion)
+
+Return an iterator over all compositions of a non-negative integer `n`.
+
+By a composition of `n` we mean a sequence of positive integers whose sum is `n`.
+
+# Examples
+```jldoctest
+julia> C = compositions(4)
+Iterator over the compositions of 4
+
+julia> collect(C)
+8-element Vector{Oscar.Composition{Int64}}:
```

Comment:
The type `Partition` is exported and thus prints as `Vector{Partition{Int64}}` in this context. Maybe export `Combination` and `WeakCombination` as well?

Reply:
Yes, why not.